### ISSUE TITLE
File locking error when resuming a h-repex simulation on multiple GPUs

### ISSUE THREAD

**Issue by zhang-ivy** [choderalab/perses#806]
I was trying to resume a single repex simulation (using 4 GPUs) with the following code:
```python
reporter = MultiStateReporter(reporter_file, checkpoint_interval=10)
simulation = HybridRepexSampler.from_storage(reporter)
# Determine how many more iterations are needed
total_iterations = 5000
iterations =  total_iterations - simulation.iteration
# Resume simulation
simulation.extend(n_iterations=iterations)
```
However, I got the following error:
```
OSError: [Errno -101] NetCDF: HDF error: b'/data/chodera/zhangi/perses_benchmark/repe
x/31/7/0/0_complex.nc'
CRITICAL:mpiplus.mpiplus:MPI node 1/4 raised an exception and called Abort()! The exc
eption traceback follows
```

This problem was found (and solved) in Yank as well: https://github.com/choderalab/yank/issues/1165

To fix this, I added the following line to my bash script: `export HDF5_USE_FILE_LOCKING=FALSE`

@jchodera suggests we fix this in a similar way to how Yank fixes this: https://github.com/choderalab/yank/pull/1168

However, the Yank code containing the fixes above has already been ported to openmmtools (see [here](https://github.com/choderalab/openmmtools/blob/master/openmmtools/multistate/multistatereporter.py#L379-L380)), so I'm not sure what the appropriate fix for perses is (or why [the lines in openmmtools](https://github.com/choderalab/openmmtools/blob/master/openmmtools/multistate/multistatereporter.py#L379-L380) aren't sufficient).

**Comment by mikemhenry** [comment]
Can you post the full traceback?

**Comment by mikemhenry** [comment]
Also something we can investigate is to use the HDF5 SWMR (single writer, multiple reader) feature which should prevent these kinds of errors
https://docs.h5py.org/en/stable/swmr.html

**Comment by zhang-ivy** [comment]
Full traceback:
```
******************************************
DEBUG:openmmtools.multistate.multistatesampler:Iteration 3421/5000
DEBUG:openmmtools.multistate.multistatesampler:**************************************
******************************************
DEBUG:mpiplus.mpiplus:Node 4/4: waiting for broadcast of <function ReplicaExchangeSam
pler._mix_replicas at 0x2b09ee38fe50>
DEBUG:mpiplus.mpiplus:Node 1/4: executing <bound method MultiStateReporter.open of <o
penmmtools.multistate.multistatereporter.MultiStateReporter object at 0x2b5e2850a280>
>
Traceback (most recent call last):
  File "/home/zhangi/choderalab/perses_benchmark/perses_protein_mutations/code/24_rbd
_ace2_repex/resume_h_repex.py", line 20, in <module>
    simulation = HybridRepexSampler.from_storage(reporter)
  File "/home/zhangi/miniconda3/envs/perses-rbd-ace2/lib/python3.8/site-packages/open
mmtools/multistate/multistatesampler.py", line 232, in from_storage
    mpiplus.run_single_node(0, sampler._reporter.open, mode='a',
  File "/home/zhangi/miniconda3/envs/perses-rbd-ace2/lib/python3.8/site-packages/mpip
lus/mpiplus.py", line 220, in run_single_node
    result = task(*args, **kwargs)
  File "/home/zhangi/miniconda3/envs/perses-rbd-ace2/lib/python3.8/site-packages/open
mmtools/multistate/multistatereporter.py", line 274, in open
    self._storage_analysis = self._open_dataset_robustly(self._storage_analysis_file_
path,
  File "/home/zhangi/miniconda3/envs/perses-rbd-ace2/lib/python3.8/site-packages/open
mmtools/multistate/multistatereporter.py", line 371, in _open_dataset_robustly
    raise e
  File "/home/zhangi/miniconda3/envs/perses-rbd-ace2/lib/python3.8/site-packages/open
mmtools/multistate/multistatereporter.py", line 364, in _open_dataset_robustly
    return netcdf.Dataset(*args, **kwargs)
  File "src/netCDF4/_netCDF4.pyx", line 2330, in netCDF4._netCDF4.Dataset.__init__
  File "src/netCDF4/_netCDF4.pyx", line 1948, in netCDF4._netCDF4._ensure_nc_success
OSError: [Errno -101] NetCDF: HDF error: b'/data/chodera/zhangi/perses_benchmark/repe
x/31/7/0/0_complex.nc'
CRITICAL:mpiplus.mpiplus:MPI node 1/4 raised an exception and called Abort()! The exc
eption traceback follows
```

**Comment by mikemhenry** [comment]
It looks like something is wrong with the logic in ` _open_dataset_robustly` since the line that gets thrown is https://github.com/choderalab/openmmtools/blob/master/openmmtools/multistate/multistatereporter.py#L371 which is before https://github.com/choderalab/openmmtools/blob/master/openmmtools/multistate/multistatereporter.py#L380 I'm pretty sure that is what the error is since if it had a chance to set the envar, it would have thrown an error here https://github.com/choderalab/openmmtools/blob/master/openmmtools/multistate/multistatereporter.py#L383

**Comment by jchodera** [comment]
Any chance you can figure out a fix for that method in `openmmtools`, @mikemhenry ?

**Comment by mikemhenry** [comment]
I will take a look! The hard part might be testing. @zhang-ivy was this bug transient or did it happen every time? If it happened every time, can you send me everything needed to reproduce?

**Comment by zhang-ivy** [comment]
It happened every time when I saw that error. 
Here are the files needed to reproduce the error:
Note that you'll need to install `mpi4py`.
[debug_file_locking.zip](https://github.com/choderalab/perses/files/6703320/debug_file_locking.zip)
The checkpoint file is too large to upload, but you can grab one here: `/data/chodera/zhangi/perses_benchmark/repex/31/3/0/0_apo.nc`

**Comment by jchodera** [comment]
Maybe we can make the suspected fix and park an issue in `openmmtools` to develop a minimal test that just exercises components in `openmmtools`? I think it could be a bit of a timesink to develop a reliable test, while the fix itself may be quick.

**Comment by mikemhenry** [comment]
The issue is in python 3.3+ `IOError` was merged into `OSError` https://docs.python.org/3/library/exceptions.html#OSError
IE
``` python
>>> try:
...  raise OSError
... except (IOError):
...  print("oh no")
... 
oh no
```
So the issue is that instead of just trying a few times, we hit the code path that assumes the file doesn't exit and exits early. Looking at the source code (and docs) of netcdf https://github.com/Unidata/netcdf4-python/blob/master/src/netCDF4/_netCDF4.pyx#L1939 It it isn't obvious which error will be thrown if the file doesn't exit verses some issue with HDF locking, so what I will do is remove the logic that handles the case of the file not exiting since the worse case scenario is instead of failing right away, it will fail in ~ 10 seconds.

**Comment by mikemhenry** [comment]
PR here: https://github.com/choderalab/openmmtools/pull/515

**Comment by jchodera** [comment]
Good catch!

**Comment by mikemhenry** [comment]
Will take a new release but fixed with choderalab/openmmtools#515
